#Add comments in each one of your code cells
##Include markdown cells in between your code cells to add titles/information to each exercise
#Indications about when to commit and push changes are included, but you are encouraged to commit and push more often.
#You won’t need to upload any data.



Title: Snowshoe hare physical data in Bonanza Creek Experimental Forest: 1999-Present
Creators:Kielland, Knut; Senior Investigator Chapin, F. Stuart; Primary Investigator Ruess, Roger W.; Primary Investigator
Location: Bonanza Creek LTER
Publication Date: 2017-02-10
Citation:
Kielland, K., F.S. Chapin, R.W. Ruess, and Bonanza Creek LTER. 2017. Snowshoe hare physical data in Bonanza Creek Experimental Forest: 1999-Present ver 22. Environmental Data Initiative. https://doi.org/10.6073/pasta/03dce4856d79b91557d8e6ce2cbcdc14 (Accessed 2023-10-19).

Abstract:
Snowshoe hares, Lepus americanus, are a 'keystone' prey species in northern boreal forests and experience population fluctuations of 8-11-years. Despite intense responses of both vegetation and predators to changes in hare densities, landscape-scale comparisons of hare populations in Alaska have been limited to qualitative descriptions. We conducted capture-recapture studies of snowshoe hares at 5 locales in the Tanana valley, from Tok in the east to Clear in the west from 1999 to 2002. Snowshoe hare densities were highest in 1999 ( =6.36 ha-1, SE=0.63) and declined thereafter. We were unable to detect declines in apparent survival during declining densities in our study populations. Movement distances did not vary temporally and persistence of individuals through declining densities may be associated positively with body condition at the peak. The relationship of hare pellets and hare densities was weak and limits the utility of this methodology for estimating hare densities in Interior Alaska.

![quoddy head hare](https://upload.wikimedia.org/wikipedia/commons/8/8a/SNOWSHOE_HARE_%28Lepus_americanus%29_%285-28-2015%29_quoddy_head%2C_washington_co%2C_maine_-01_%2818988734889%29.jpg)

In [12]:
import pandas as pd 
hares = pd.read_csv("https://portal.edirepository.org/nis/dataviewer?packageid=knb-lter-bnz.55.22&entityid=f01f5d71be949b8c700b6ecd1c42c701")

In [13]:
hares.head()

date time    grid trap       l_ear r_ear  sex  age  weight  hindft  \
0  11/26/1998  NaN  bonrip   1A  414D096A08   NaN  NaN  NaN  1370.0   160.0   
1  11/26/1998  NaN  bonrip   2C  414D320671   NaN    M  NaN  1430.0     NaN   
2  11/26/1998  NaN  bonrip   2D  414D103E3A   NaN    M  NaN  1430.0     NaN   
3  11/26/1998  NaN  bonrip   2E  414D262D43   NaN  NaN  NaN  1490.0   135.0   
4  11/26/1998  NaN  bonrip   3B  414D2B4B58   NaN  NaN  NaN  1710.0   150.0   

  notes  b_key  session_id       study  
0   NaN  917.0          51  Population  
1   NaN  936.0          51  Population  
2   NaN  921.0          51  Population  
3   NaN  931.0          51  Population  
4   NaN  933.0          51  Population

In [26]:
hares_og = hares.sex.value_counts()
hares.sex.value_counts()

F     1161
M      730
f      556
m      515
?       40
F?      10
f        4
m        4
f?       3
M?       2
m?       2
pf       1
Name: sex, dtype: int64

In [27]:
hares_dropna = hares.sex.value_counts(dropna=False)
hares.sex.value_counts(dropna=False)

F      1161
M       730
f       556
m       515
NaN     352
?        40
F?       10
f         4
m         4
f?        3
M?        2
m?        2
pf        1
Name: sex, dtype: int64

The difference appears to be the addition of the NaN row. Dropna = False added an additional row that specified the number of NA observation (352). 

In [23]:
pd.unique(hares_og)

array([1161,  730,  556,  515,   40,   10,    4,    3,    2,    1])

In [30]:
hares_dropna.unique()

array([1161,  730,  556,  515,  352,   40,   10,    4,    3,    2,    1])

You likely noticed there seem to be some repeated values, for example m appears twice. Use the unique() method on the sex column to see the unique non-NA values in this column. Discuss with your team what was the cause of the seemingly repeated values.

In [31]:
hares.sex.unique()

array([nan, 'M', 'F', '?', 'F?', 'M?', 'pf', 'm', 'f', 'f?', 'm?', 'f ',
       'm '], dtype=object)

The ones that are allowed are: M, F, or M? , and F?

In [57]:
import numpy as np

conditions = [(hares.sex=='F'), (hares.sex=='f'), (hares.sex=='M') ,(hares.sex=='m'), (hares.sex=='F?'),(hares.sex=='M?')]
choicelist = ['female', 'female', 'male', 'male', 'female', 'male']
np.select(conditions, choicelist, default=np.nan)


array(['nan', 'male', 'male', ..., 'nan', 'nan', 'male'], dtype='<U32')

In [55]:
len(conditions)

6